core/src/serde/de.rs

```diff
@@ -0,0 +1,693 @@
+    }
+
+    fn deserialize_newtype_struct<V: de::Visitor<'de>>(
+        self,
+        name: &str,
+        visitor: V,
+    ) -> Result<V::Value, Self::Error> {
+        if name == CID_SERDE_PRIVATE_IDENTIFIER {
+            match self {
+                Ipld::Link(cid) => visitor.visit_newtype_struct(CidDeserializer(cid)),
+                _ => error(format!(
+                    "Only `Ipld::Link`s can be deserialized to CIDs, input was `{:#?}`",
+                    self
+                )),
+            }
+        } else {
+            visitor.visit_newtype_struct(self)
+        }
+    }
+
+    // Heavily based on
+    // https://github.com/serde-rs/json/blob/95f67a09399d546d9ecadeb747a845a77ff309b2/src/value/de.rs#L249
+    fn deserialize_enum<V: de::Visitor<'de>>(
+        self,
+        _name: &str,
+        _variants: &[&str],
+        visitor: V,
+    ) -> Result<V::Value, Self::Error> {
+        let (variant, value) = match self {
+            Ipld::Map(map) => {
+                let mut iter = map.into_iter();
+                let (variant, value) = match iter.next() {
+                    Some(v) => v,
+                    None => {
+                        return error(
+                            "Only `Ipld::Map`s with a single key can be deserialized to `enum`, input had no keys"
+                        );
+                    }
+                };
+                // Enums are encoded in IPLD as maps with a single key-value pair
+                if iter.next().is_some() {
+                    return error(
+                        "Only `Ipld::Map`s with a single key can be deserialized to `enum`, input had more keys"
+                    );
+                }
+                (variant, Some(value))
+            }
+            Ipld::String(variant) => (variant, None),
+            _ => return error(format!(
+                    "Only `Ipld::Map` and `Ipld::String` can be deserialized to `enum`, input was `{:#?}`",
+                    self
+            )),
+        };
+
+        visitor.visit_enum(EnumDeserializer { variant, value })
+    }
+
+    // Heavily based on
+    // https://github.com/serde-rs/json/blob/95f67a09399d546d9ecadeb747a845a77ff309b2/src/value/de.rs#L446
+    fn deserialize_ignored_any<V: de::Visitor<'de>>(
+        self,
+        visitor: V,
+    ) -> Result<V::Value, Self::Error> {
+        drop(self);
+        visitor.visit_none()
+    }
+
+    fn deserialize_option<V: de::Visitor<'de>>(self, visitor: V) -> Result<V::Value, Self::Error> {
+        match self {
+            Self::Null => visitor.visit_none(),
+            _ => visitor.visit_some(self),
+        }
+    }
+}
+
+fn visit_map<'de, V>(map: BTreeMap<String, Ipld>, visitor: V) -> Result<V::Value, SerdeError>
+where
+    V: de::Visitor<'de>,
+{
+    let mut deserializer = MapDeserializer::new(map);
+    visitor.visit_map(&mut deserializer)
+}
+
+fn visit_seq<'de, V>(list: Vec<Ipld>, visitor: V) -> Result<V::Value, SerdeError>
```

Comment:
We're dropping the size hints here.

Reply:
Oh, never mind. I noticed we were dropping `_len` in https://github.com/ipld/libipld/blob/ad768fe128a76c2c1ea0fcb40bbcc1ff23644ae8/core/src/serde/de.rs#L696, but we don't actually need that (the list tells us what we need).